core/shared/src/main/scala/zio/Schedule2.scala

```diff
@@ -0,0 +1,576 @@
+/*
+ * Copyright 2017-2020 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio
+
+import java.time.Instant
+import scala.util.Try
+
+import zio.duration._
+import java.util.concurrent.TimeUnit
+
+/**
+ * A `Schedule[Env, In, Out]` defines a recurring schedule, which consumes values of type `In`, and
+ * which returns values of type `Out`.
+ *
+ * Schedules are defined as a possibly infinite set of intervals spread out over time. Each
+ * interval defines a window in which recurrance is possible.
+ *
+ * When schedules are used to repeat or retry effects, the starting boundary of each interval
+ * produced by a schedule is used as the moment when the effect will be executed again.
+ *
+ * Schedules compose in the following primary ways:
+ *
+ *  * Union. This performs the union of the intervals of two schedules.
+ *  * Intersection. This performs the intersection of the intervals of two schedules.
+ *  * Sequence. This concatenates the intervals of one schedule onto another.
+ *
+ * In addition, schedule inputs and outputs can be transformed, filtered (to  terminate a
+ * schedule early in response to some input or output), and so forth.
+ *
+ * A variety of other operators exist for transforming and combining schedules, and the companion
+ * object for `Schedule` contains all common types of schedules, both for performing retrying, as
+ * well as performing repetition.
+ */
+final case class Schedule2[-Env, -In, +Out](
+  private val step0: Schedule2.StepFunction[Env, In, Out]
+) { self =>
+  import Schedule2._
+  import Schedule2.Decision._
+
+  def ||[Env1 <: Env, In1 <: In, Out2](that: Schedule2[Env1, In1, Out2]): Schedule2[Env1, In1, (Out, Out2)] =
```

Comment:
This isn't really a comment about the code, just a question about code naming convention.  Why refer to it as `Env1`/`In1` instead of `Env2`/`In2`.  But then use `Out2` instead of `Out1`?

Reply:
I use `Env1` when it's a 'variation' on `Env` (I'd really prefer to use `Env'`, but it's not possible!). I use `Env2` when it is a different environment with no relation to `Env`.

Reply:
Thanks!